### DISCUSSION TITLE
Make AppThemeBinding type public

### DISCUSSION THREAD

**Discussion by egvijayanand** [dotnet/maui#22250] [Ideas]
### Description

While working with C# Markup, the type `AppThemeBinding` is the equivalent of `AppThemeBindingExtension` from XAML.

But this type is marked as internal. So, unable to make use of it while defining theme-based Resource definition (as an instance assigned to the value of the Setter).

_For inline View definition, there's an extension method available. But not for Setter as there's no instance associated to invoke the extension method. Refer to the intended use case for the sample._

### Public API Changes

Make the **AppThemeBinding** type public.

https://github.com/dotnet/maui/blob/c3221bf08d291c9f69070aa11d6d60f61a569aae/src/Controls/src/Core/AppThemeBinding.cs#L9

### Intended Use-Case

Allows resource reuse.

```cs
Resources.Add("MauiLabel", new Style(typeof(Label))
{
    Setters =
    {
        new() { Property = Label.TextColorProperty, Value = new AppThemeBinding() { Dark = Colors.White, Light = Colors.Black } },
    },
});
```

**Comment by pictos** [comment]
@brminnick C# markup does have access to internal methods, right? If so, can we create a Factory method for `AppThemeBinding` ? Or is there something already?

**Comment by egvijayanand** [comment]
> @brminnick C# markup does have access to internal methods, right? If so, can we create a Factory method for `AppThemeBinding` ? Or is there something already?

Yeah, it can even be in the Markup package as there's already an extension method available for AppThemeBinding (instance-based) but not for **Setter**. The return type has to be `BindingBase` and invoked directly.

**Comment by TheCodeTraveler** [comment]
We have extension methods, [`.AppThemeBinding()` and `.AppThemeColorBinding()`](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/markup/theming#1-on-a-specific-bindableobject)

```cs
new Label()
    .AppThemeBinding(Label.TextProperty, "Light Mode", "Dark Mode")
    .AppThemeColorBinding(Label.TextColorProperty, Colors.Black, Colors.White);

var labelStyle = new Style<Label>()
    .AddAppThemeBinding(Label.TextProperty, "Light Mode", "Dark Mode")
    .AddAppThemeBinding(Label.TextColorProperty, Colors.Black, Colors.White);

new Label
{
    Style = labelStyle
};

```

**Comment by egvijayanand** [comment]
> We have an extension method, [`.AppThemeBinding()`](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/markup/extensions/bindable-object-extensions#appthemebinding)

Yes, but won't work in the context of **Setter** as there's no instance available to invoke this form of method. Style can be implicit also.

Maybe will try with the another version updated in the earlier comment.

**Comment by egvijayanand** [comment]
@brminnick, it works seamlessly with the generic version `Style<T>`.

But one request, can another extension method be added to make it work with the non-generic version of Style to support the existing code base?

```cs
// Since Style is sealed, have defined as it is
public static Style AddAppThemeBinding(this Style style, BindableProperty property, object light, object dark)
{
    // Issue is AppThemeBinding is inaccessible due to its protection level
    style.Setters.Add(property, new AppThemeBinding() { Light = light, Dark = dark });
    return style;
}
```

**Comment by egvijayanand** [comment]
@brminnick Or to make it simple, add another constructor to generic `Style<T>` that takes a non-generic `Style` as a parameter. Any further operation after this is the same as working with that Style instance.

Since `Style<T>` can be implicitly converted to `Style`, the other way is not all an issue.

```cs
// Adding a condition to make it even more safer
public Style(Style style)
{
    // Assign to the underlying property only if the Style's TargetType matches with that of the generic Type 'T'
    if (typeof(T) == style.TargetType)
    {
        MauiStyle = style;
    }
    else
    {
        MauiStyle = new Style(typeof(T));
    }
}
```

**Comment by TheCodeTraveler** [comment]
@egvijayanand I think that makes a lot of sense! 

Check out the proposal I just opened and leave a comment with your thoughts! 
https://github.com/CommunityToolkit/Maui.Markup/issues/274

**Comment by Ghostbird** [comment]
I found out that I can't access the `AppThemeBinding`. How do I make a multi-binding on text colour that also adapts to theme changes?

It makes no sense that MAUI exists, but you need to install a community library to be able to bind to the app theme changes at all. And the community library still does not expose the `AppThemeBinding`.

You might say: _But I can already bind to the AppTheme in XAML_. Exactly! Why is it publicly accessible in one language (XAM‌L) but not in C#? It is extremely weird.

**Comment by egvijayanand** [comment]
This remains a valid request as I encounter issues defining the Visual States. 

Currently, no alternative method, even with the CommunityToolkit, which defines the `AppThemeBinding` extension method at the Style level rather than at the Setter level to work with Visual States.

I don't understand why there is a disparity in non-XAML development.

**Comment by stephenquan** [comment]
With regards to the Style, I believe @brminnick already answers this in his post https://github.com/dotnet/maui/discussions/22250#discussioncomment-9338233 which describes [.AppThemeColorBinding()](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/markup/theming#1-on-a-specific-bindableobject) when applied to your example becomes:

```c#
Resources.Add("MauiLabel", new Style<Label>()
    .AddAppThemeBinding(Label.TextColorProperty, Colors.Black, Colors.White));
```

**Reply by egvijayanand** [reply]
That extension method is defined for the Style type, whereas my concern is for Visual State, which is defined at the individual Setter instance. This is still not possible.

Here's a sample from the official template. Line # 59 & 60

https://github.com/dotnet/maui/blob/5b807868ad2fb36c0794853afedffc3deca34610/src/Templates/src/templates/maui-mobile/Resources/Styles/Styles.xaml#L53-L66

**Comment by stephenquan** [comment]
@egvijayanand, here's a bunch of classes you can use to improve your C# app theme binding experience:
 - AppThemeEx - a wrapper to RequestedTheme and UserAppTheme with PropertyChanged notification
 - AppThemeExConverter - a MultiValueConverter which takes 3 inputs (RequestedTheme, lightObject, darkObject) and returns lightObject or darkObject based on input
 - AppThemeExBinding - a wrapper of the above two classes to produce a BindingBase, the light/dark inputs can be objects or bindings to objects

Here's a sample usage:

```c#
boxView.SetBinding(BoxView.ColorProperty, AppThemeExBinding.Create(Colors.Green, Colors.Red));
```

The advantage of this implementation is that you can use `AppThemeExBinding.Create()` in various scenarios, including your style setter scenarios. Here are the classes:

```c#
public partial class AppThemeEx : INotifyPropertyChanged
{
    static AppThemeEx? instance;
    public static AppThemeEx Current => instance ??= new AppThemeEx();

    AppTheme? requestedTheme;
    public AppTheme? RequestedTheme => Application.Current?.RequestedTheme;

    AppTheme userAppTheme = AppTheme.Unspecified;
    public AppTheme UserAppTheme
    {
        get => userAppTheme;
        set
        {
            userAppTheme = value;
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(UserAppTheme)));
            if (Application.Current is Application app)
            {
                app.UserAppTheme = value;
            }
            Poll();
        }
    }

    public void Poll()
    {
        if (Application.Current is Application app)
        {
            if (requestedTheme != app.RequestedTheme)
            {
                requestedTheme = app.RequestedTheme;
                PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(RequestedTheme)));
            }
            if (userAppTheme != app.UserAppTheme)
            {
                userAppTheme = app.UserAppTheme;
                PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(UserAppTheme)));
            }
        }
    }

    public AppThemeEx()
    {
        if (Application.Current is Application app)
        {
            app.RequestedThemeChanged += (s, e) => Poll();
        }
    }

    public event PropertyChangedEventHandler? PropertyChanged;
}


public class AppThemeExConverter : IMultiValueConverter
{
    public object? Convert(object?[] values, Type targetType, object? parameter, CultureInfo culture)
    {
        if (values.Length >= 3 && values[0] is AppTheme theme)
        {
            switch (values[0])
            {
                case AppTheme.Light:
                    return values[1];
                case AppTheme.Dark:
                    return values[2];
            }
            return Application.Current?.RequestedTheme != AppTheme.Dark ? values[1] : values[2];
        }
        return null;
    }

    public object[] ConvertBack(object value, Type[] targetTypes, object? parameter, CultureInfo culture)
    {
        throw new NotImplementedException();
    }
}

public static class AppThemeExBinding
{
    public static BindingBase Create(object? light = null, object? dark = null)
    {
        BindingBase appThemeBinding = BindingBase.Create(static (AppThemeEx ate) => ate.RequestedTheme, BindingMode.OneWay, source: AppThemeEx.Current);
        BindingBase lightBinding = light is BindingBase _lightBinding ? _lightBinding : BindingBase.Create(static (object? o) => o, BindingMode.OneWay, source: light);
        BindingBase darkBinding = dark is BindingBase _darkBinding ? _darkBinding : BindingBase.Create(static (object? o) => o, BindingMode.OneWay, source: dark);
        return new MultiBinding
        {
            Bindings = new List<BindingBase>
            {
                appThemeBinding,
                lightBinding,
                darkBinding
            },
            Converter = new AppThemeExConverter()
        };
    }
}
```

**Comment by ahoke-cr** [comment]
This seems to have stalled out, but I've encountered a scenario I'm not sure how to address. I'm using compiled bindings which make data type matches apparently vital to bindings being correctly handled. I use FFImageLoading to render SVGs in varying colors based on app theme, icon purpose (warning, error, info, etc.), or user customization using the ReplaceStringMap bindable property. It outputs a dictionary of strings to replace in the SVG definition (which I replace fill colors using a replace string like `fill="#000000"` to replace the default SVG color with the desired color.

This replace map is generated by passing a Color into a converter that generates the dictionary, assigned using a Binding with the source being the color (often an AppThemeBindingExtension to have icons that responsively change colors to match the desired system theme). Now, the binding context and inherited `x:DataType` of the SVG is not AppThemeBinding and I get a slew of binding failures which actually result in the binding failing completely in many cases. I've found I'm able to resolve this by setting the `x:DataType` of the Binding markup extension to `{x:Type x:AppThemeBindingExtension}` but this is not the actual output type of the extension, which is this internal type under discussion: `AppThemeBinding`:
<img width="1293" height="115" alt="image" src="https://github.com/user-attachments/assets/898d426b-0e92-4475-a933-6acbe7292605" />
